src/functional/fencil_processors/processor_interface.py

```diff
@@ -0,0 +1,139 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+"""
+Interface for fencil processors.
+
+Fencil processors are functions which take an ``iterator.ir.FencilDefinition``
+along with input values for the fencil. Those which execute the fencil with
+the given arguments (possibly by generating code along the way) are fencil
+executors. Those which generate any kind of string based on the fencil
+and (optionally) input values are fencil formatters.
+
+For more information refer to
+``gt4py/docs/functional/architecture/007-Fencil-Processors.md``
+"""
+from __future__ import annotations
+
+from typing import Protocol, TypeGuard, TypeVar, cast
+
+from functional.iterator.ir import FencilDefinition
+
+from .source_modules import SourceModule
+
+
+OutputT = TypeVar("OutputT", covariant=True)
+ProcessorKindT = TypeVar("ProcessorKindT", bound="FencilProcessorProtocol", covariant=True)
+
+
+class FencilProcessorFunction(Protocol[OutputT]):
+    def __call__(self, fencil: FencilDefinition, *args, **kwargs) -> OutputT:
+        ...
+
+
+class FencilProcessorProtocol(FencilProcessorFunction[OutputT], Protocol[OutputT, ProcessorKindT]):
+    @property
+    def kind(self) -> type[ProcessorKindT]:
+        ...
+
+
+class FencilFormatter(FencilProcessorProtocol[str, "FencilFormatter"], Protocol):
+    @property
+    def kind(self) -> type[FencilFormatter]:
+        return FencilFormatter
+
+
+def fencil_formatter(
+    func: FencilProcessorFunction[str],
+) -> FencilProcessorProtocol[str, FencilFormatter]:
```

Comment:
Doesn't mypy accept the concrete `FencilFormatter` type as return type?

```suggestion
def fencil_formatter(func: FencilProcessorFunction[str]) -> FencilFormatter:
```

Reply:
It does not consider `FencilFormatter` a concrete type. That makes sense, because it is deriving from `Protocol`.

Reply:
Basically the two are exchangeable here in every regard.

Reply:
Great. Could you also remove the unnecessary comma after the `func` parameter so black will be allowed to put the signature in one single line?